File: app/src/main/java/com/jodelXposed/krokofant/utils/Log.java
package com.jodelXposed.krokofant.utils;

public class Log {
    public static void xlog(String s) {
        String TAG;
        StackTraceElement[] stElements = Thread.currentThread().getStackTrace();
        TAG = String.format("%s@%s JodelXposed",
            stElements[3].getClassName(),
            stElements[3].getMethodName());
        try {
            Class.forName("de.robv.android.xposed.XposedBridge", false, Log.class.getClassLoader());
            de.robv.android.xposed.XposedBridge.log(TAG + ": " + s);
        } catch (ClassNotFoundException e) {
            android.util.Log.i(TAG, s);
        }
    }
}

File: app/src/main/java/com/jodelXposed/utils/Log.kt
package com.jodelXposed.utils

import de.robv.android.xposed.XposedBridge

object Log {
    @JvmStatic
    val TAG: String = "JodelXposed"

    @JvmStatic
    fun xlog(msg: String?) {
        try {
            Class.forName("de.robv.android.xposed.XposedBridge", false, Log::class.java.classLoader)
            XposedBridge.log("$TAG: $msg")
        } catch (e: ClassNotFoundException) {
            android.util.Log.e(TAG, "Tried to output to Xposed log: $msg")
        }
    }

    @JvmStatic
    fun xlog(t: Throwable?) {
        try {
            Class.forName("de.robv.android.xposed.XposedBridge", false, Log::class.java.classLoader)
            XposedBridge.log(t)
        } catch (e: ClassNotFoundException) {
            android.util.Log.e(TAG, "Tried to output to Xposed log", t)
        }
    }

    @JvmStatic
    fun xlog(msg: String?, t: Throwable?) {
        xlog(msg)
        xlog(t)
    }

    @JvmStatic
    fun dlog(msg: String?, t: Throwable? = null) {
        if (t == null)
            android.util.Log.d(TAG, msg)
        else
            android.util.Log.d(TAG, msg, t)
    }

    @JvmStatic
    fun dlog(msg: String?) = dlog(msg, null)
}
